File: dynamic_programing/Javascript/Kadane.js
/**
 * Kadane's algorithm is one of the most efficient ways to
 * calculate the maximum contiguos subarray sum for a given array.
 * Below is the implementation of kadanes's algorithm along with
 * some sample test cases.
 * There might be a special case in this problem if al the elements
 * of the given array are negative. In such a case, the maximum negative
 * value present in the array is the answer.
 */

function kadaneAlgo(array) {
  let cummulativeSum = 0;
  let maxSum = Number.NEGATIVE_INFINITY; // maxSum has the least posible value
  for (let i = 0; i < array.length; i++) {
    cummulativeSum = cummulativeSum + array[i];
    if (maxSum < cummulativeSum) {
      maxSum = cummulativeSum;
    } else if (cummulativeSum < 0) {
      cummulativeSum = 0;
    }
  }
  return maxSum;
  // This function returns largest sum contiguous sum in a array
}

function main() {
  let arr = [1, 2, 3, -5, -6];
  let sum = kadaneAlgo(arr);
  console.log(sum);
}
main();

File: dynamic_programing/C++/kadane.cpp
#include <climits>  
#include <iostream> 
using namespace std;

int max_subarray_sum(int arr[], int length)
{
    int current_max = INT_MIN, current_sum = 0;
    for (int i = 0; i < length; i++)
    {
        current_sum = current_sum + arr[i];
        if (current_max < current_sum)
        {
            current_max = current_sum;
        }

        if (current_sum < 0)
        {
            current_sum = 0;
        }
    }
    return current_max;
}
int main()
{
    int n; 
    cout << "Enter length of the array: ";
    cin >> n;
    int arr[n]; 

    for (int i = 0; i < n; i++) 
    {
            cin >> arr[i];
    }
    int max_sum = max_subarray_sum(arr, n); 
    cout << "Maximum contiguous sum for this array is : " << max_sum << endl;

    return 0;
}
